misc/multiaddr/src/lib.rs

```diff
@@ -73,9 +70,10 @@ impl Multiaddr {
     /// ```
     ///
     pub fn push(&mut self, p: Protocol<'_>) {
-        let mut w = io::Cursor::<&mut Vec<u8>>::new(Arc::make_mut(&mut self.bytes));
+        let mut w = io::Cursor::new(self.to_vec());
```

Comment:
Can we not avoid this allocation if `p` fits into `Storage::Inline`?

Reply:
I guess so, but it would be a bit more tricky, and I was going for a minimal code change.

If you write a thing that behaves like a vector and gracefully switches from inline to heap storage, at some point you are basically reimplementing https://crates.io/crates/smallvec . You could just have a dependency on smallvec, then this would be super easy to write efficiently. But I have chosen to go with fewer dependencies.

I doubt that it matters, since typically you build or deconstruct a multiaddr once, but then keep it around for a long time. There are lots of multiaddrs that you *never* build but just get over the wire somehow.

Reply:
> You could just have a dependency on smallvec, then this would be super easy to write efficiently. But I have chosen to go with fewer dependencies.

I would go with `smallvec` which is already used in quite a few places within `rust-libp2p`.

Reply:
Using smallvec now.